src/Bicep.LangServer/Handlers/BicepDecompileCommandHandler.cs

```diff
@@ -0,0 +1,191 @@
+        }
+
+        public BicepDecompileCommandResult(
+            string output,
+            string? errorMessage
+        ) : this(output, errorMessage, null, new DecompiledFile[] { }, new string[] { })
+        {
+        }
+
+        public BicepDecompileCommandResult(
+            string output,
+            string mainBicepPath,
+            DecompiledFile[] outputFiles,
+            string[] conflictingOutputPaths
+        ) : this(output, null, mainBicepPath, outputFiles, conflictingOutputPaths)
+        {
+        }
+    }
+
+    /// <summary>
+    /// Handles a request from the client to decompile a JSON file for given a file path, creating one or more bicep files
+    /// </summary>
+    public class BicepDecompileCommandHandler : ExecuteTypedResponseCommandHandlerBase<BicepDecompileCommandParams, BicepDecompileCommandResult>
+    {
+        private readonly TemplateDecompiler templateDecompiler;
+        private readonly TelemetryAndErrorHandlingHelper<BicepDecompileCommandResult> telemetryHelper;
+
+        public BicepDecompileCommandHandler(
+            ISerializer serializer,
+            IFeatureProviderFactory featureProviderFactory,
+            INamespaceProvider namespaceProvider,
+            IModuleRegistryProvider registryProvider,
+            ILanguageServerFacade server,
+            ITelemetryProvider telemetryProvider,
+            IApiVersionProviderFactory apiVersionProviderFactory,
+            IBicepAnalyzer bicepAnalyzer,
+            IFileResolver fileResolver)
+            : base(LangServerConstants.DecompileCommand, serializer)
+        {
+            this.telemetryHelper = new TelemetryAndErrorHandlingHelper<BicepDecompileCommandResult>(server.Window, telemetryProvider);
+            this.templateDecompiler = new TemplateDecompiler(featureProviderFactory, namespaceProvider, fileResolver, registryProvider, apiVersionProviderFactory, bicepAnalyzer);
+        }
+
+        public override Task<BicepDecompileCommandResult> Handle(BicepDecompileCommandParams parameters, CancellationToken cancellationToken)
+        {
+            return telemetryHelper.ExecuteWithTelemetryAndErrorHandling(() =>
+            {
+                if (parameters.jsonUri is null)
+                {
+                    throw new ArgumentNullException(nameof(parameters.jsonUri));
+                }
+
+                return Task.FromResult(Decompile(parameters.jsonUri.GetFileSystemPath()));
+            });
+        }
+
+        private (BicepDecompileCommandResult result, BicepTelemetryEvent? successTelemetry) Decompile(string jsonPath)
+        {
+            StringBuilder output = new StringBuilder();
+
+            Uri jsonUri = new Uri(jsonPath, UriKind.Absolute);
+
+            Uri? bicepUri;
+            ImmutableDictionary<Uri, string>? filesToSave;
+            try
+            {
+                // Decompile
+                Log(output, $"Decompiling {jsonPath} into Bicep...");
```

Comment:
Shouldn't this be localized?

Reply:
I wish we would either require it or not worry about it.

Reply:
972cee6e7